### PR TITLE
Issue #34: Browser Class Activation Maps for the Swin classifier

### PR DESCRIPTION
**Pull request by danilo-bukvic**
## What this adds

Model explainability for the Swin classifier (issue #34): **Class Activation Maps (CAM) that show which image regions drive each species prediction**, computed entirely in the browser. No backend changes.

Under each run in the Images panel, global **Concept 1…N** toggles appear (one per prediction rank). Turning on **Concept N** overlays every seed's own rank-N species heatmap at once — a jet (blue→red) colormap over each seed. Single-select: one concept shows at a time.

## How to see it

Pick the **`swin-L 101spp CAM`** classifier and run inference. Then flip the **Concept** toggles under the run in the Images panel. The plain `swin-L 101spp` entry is the same model without the explainability output, so its UI stays hidden.

## How it works

Swin's classification head is a global-average-pool over the 144 spatial tokens followed by a single linear layer, so a class logit is *exactly* the average of per-token contributions:

> `logit_c = mean_t( features[t] · W_c ) + b_c`

That makes the per-token term `features[t] · W_c` a **gradient-free, exact** class activation map — no Grad-CAM backprop needed. The browser reads the pre-pool token features from the model's `swin_layernorm` output and multiplies them by the classifier head weights `W`.

The stock model doesn't expose `swin_layernorm`, so the `swin-L 101spp CAM` entry points at a **patched ONNX export** (published on Hugging Face) that outputs it. The head weights `W` (101×1536, fp32) are extracted offline and **hosted on the same Hugging Face model repo**, fetched at runtime by URL — the app already downloads the model itself from HF the same way, and the file is model-specific so it belongs with the model rather than in this repo. CAM only runs when `swin_layernorm` is present, so every other classifier is completely unaffected (no UI, no errors, no download).

## Files

**New**
- `src/inference/cam.ts` — CAM math + head-weight fetch/cache
- `src/common/heatmapColors.ts` — jet colormap
- `src/components/CamRankToggles.tsx` — the Concept toggles

**Changed**
- `src/inference/worker.ts` — reads `swin_layernorm`, streams `cam-result` per seed
- `src/inference/models.ts` — `swin-L 101spp CAM` entry + HF file-URL helper
- `src/inference/useInference.ts`, `src/stores/useInferenceStore.ts` — CAM result/rank state
- `src/components/InferenceOverlay.tsx` — jet overlay canvas per box
- `src/components/ImageViewer.tsx`, `src/components/ImageGallery.tsx` — wiring
- `.gitignore`, version bump

**Tests**
- New: `cam.test.ts`, `heatmapColors.test.ts`, `CamRankToggles.test.tsx`; extended `useInferenceStore.test.ts`

## Testing

`tsc`, `eslint` (0 warnings), `prettier --check`, and `vite build` all pass. Unit + component tests pass in both the jsdom and browser (Chromium) suites. CAM math verified against the reference notebook (concepts match).

### CODE REVIEW

File: nachet-mini/src/assets/classifier_head_101spp.f32.bin
Comment (ChromaticPanic) [inline nachet-mini/src/assets/classifier_head_101spp.f32.bin:1]: We shouldn't commit binary files. It would be better for this to go up on huggingface since it is model specific

**Review by ChromaticPanic — CHANGES_REQUESTED** [review]
The new lines need test cases.

I don't see any toggles for the layers

Image

**Comment by ChromaticPanic** [comment]
> ## What this adds
> 
> Model-explainability for the Swin classifier (issue #34): **concept heatmaps that show what the model focused on for each seed**, all in the browser. Two views:
> 
>     * a colored **concept-map overlay** on each seed (toggle on/off per seed), and
> 
>     * per-seed rows in the Images panel showing **4 cutouts, one per concept**, like the GradCAM notebook.
> 
> 
> No backend changes.
> ## How to see it
> 
> Pick the new **`swin-L 101spp DFF`** classifier and run inference. The plain `swin-L 101spp` is the same model without the explainability UI.
> ## How it works
> 
> The heatmaps need a layer the model normally keeps internal (`swin.layernorm`). The stock model doesn't expose it, so we use a patched export that outputs it, and the browser reads it and does the heatmap math (DFF) client-side. The UI only appears when that layer is present — so other classifiers are completely unaffected (no UI, no errors).
> 
> The `swin-L 101spp DFF` entry points at the patched model already published on HuggingFace, so this works in production.
> ## Files (11)
> 
> `dff.ts` (new), `SeedConceptCutouts.tsx` (new), `worker.ts`, `models.ts`, `useInference.ts`, `useInferenceStore.ts`, `InferenceOverlay.tsx`, `ImageViewer.tsx`, `ResultsTable.tsx`, `ImageGallery.tsx`, `.gitignore`.
> ## Testing
> 
> `tsc`, `eslint` (0 warnings), and component tests pass. Verified against the reference notebook (concepts match).

There is no swin-L 101spp DFF in the model list

File: nachet-mini/src/inference/cam.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      for (let ch = 0; ch < channels; ch++) {
+        s += features[fOff + ch] * W[wOff + ch];
+      }
+      map[t] = s;
+      if (s < mn) mn = s;
+      if (s > mx) mx = s;
+    }
+    // min-max normalize to [0, 1] for display
+    const range = mx - mn || 1;
+    for (let t = 0; t < tokens; t++) map[t] = (map[t] - mn) / range;
```
Comment (youssefjedidi) [inline nachet-mini/src/inference/cam.ts]: Could we handle negative CAM values before normalizing? With `[-5, -3, -1]`, `-1` ends up red even though it is still lowering the score. Maybe apply ReLU first, like Grad-CAM, so red only means positive support. Could we add tests for negative and mixed values too?

File: nachet-mini/src/stores/tests/useInferenceStore.test.ts [hunk 2]
```diff
@@ -175,6 +201,23 @@ describe("useInferenceStore", () => {
       useInferenceStore.getState().removeResultsForImage(99);
       expect(useInferenceStore.getState().results.size).toBe(1);
     });
```
Comment (youssefjedidi) [inline nachet-mini/src/stores/tests/useInferenceStore.test.ts:203]: Could `removeResult()` also clear its CAM data and selected rank? Right now those stay in the store. Since box IDs are reused, I think an old heatmap could show up if the same image and model are run again. `removeResultsForImage()` already handles this, so maybe we can do the same here and add a small test.

File: nachet-mini/src/inference/cam.ts (hunk 1 above)
Comment (danilo-bukvic) [reply nachet-mini/src/inference/cam.ts]: Oh yeah you're so right min-max normalization stretches the range regardless of sign, so a still-negative contribution lands at the top and reads red. I'll ReLU the per-token contributions before normalizing (clamp negatives to 0, then divide by the max) so red only means positive support and regions that argue against the class stay blue. Adding tests for all-negative and mixed inputs too. Thanks!

File: nachet-mini/src/stores/tests/useInferenceStore.test.ts (hunk 2 above)
Comment (danilo-bukvic) [reply nachet-mini/src/stores/tests/useInferenceStore.test.ts:203]: Good catch. removeResult was only dropping the result and active key, so the box CAM maps and selected rank got left behind. Fixed it to clear those too (same prefix approach as removeResultsForImage), and added a test.

File: nachet-mini/src/inference/cam.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+/**
+ * Compute CAM heatmaps for the given class indices.
+ * @param features flat `swin_layernorm` data, layout (1, tokens, channels).
+ * @param tokens   spatial tokens (e.g. 144).
+ * @param channels feature dim (must be 1536 to match the head).
+ * @param classIndices which class rows of W to map (e.g. the top-K indices).
+ * @param headUrl  URL of the classifier head weights (hosted on Hugging Face).
+ */
```
Comment (ChromaticPanic) [inline nachet-mini/src/inference/cam.ts:66]: Maybe remove the @param annotation since we don't use that style anywhere else

File: nachet-mini/src/inference/models.ts [hunk 4]
```diff
@@ -218,6 +246,21 @@ export const huggingFaceUrl = (modelId: string): string => {
...
+  return `https://huggingface.co/${modelId}/resolve/main/${filename}`;
+};
+
+/**
+ * Filename of the extracted Swin classifier-head weights (101 × 1536 float32),
+ * hosted alongside the patched model on Hugging Face and fetched at CAM time.
+ * Model-specific, so it lives in the model repo rather than the app bundle.
+ */
+export const CLASSIFIER_HEAD_FILENAME = "classifier_head_101spp.f32.bin";
+
```
Comment (ChromaticPanic) [inline nachet-mini/src/inference/models.ts]: should this be a property of the model in the model list above instead?

**Review by ChromaticPanic — CHANGES_REQUESTED** [review]

File: nachet-mini/src/inference/cam.ts (hunk 3 above)
Comment (danilo-bukvic) [reply nachet-mini/src/inference/cam.ts:66]: Done!

File: nachet-mini/src/inference/models.ts (hunk 4 above)
Comment (danilo-bukvic) [reply nachet-mini/src/inference/models.ts]: moved it to a classifierHeadFile property on the model entry, buildModelConfig passes it through and the worker gates CAM on it.

**Comment by ChromaticPanic** [comment]
image

For CAM we we should label it with Top 1, Top 2, Top 3, etc instead of concept

**Comment by ChromaticPanic** [comment]
The last thing is make sure to use i18n for text and tooltips and have french translation keys

**Review by ChromaticPanic — APPROVED** [review]

**Review by youssefjedidi — APPROVED** [review]
